frontend/dockerfile/linter/ruleset.go

```diff
@@ -0,0 +1,110 @@
+package linter
+
+import (
+	"strings"
+
+	"github.com/moby/buildkit/frontend/dockerfile/parser"
+)
+
+const (
+	RuleStageNameCasing = "StageNameCasing"
+	RuleFlagCasing      = "FlagCasing"
+	RuleCommandCasing   = "CommandCasing"
+)
+
+var LinterRules map[string]LinterRule
+
+type LinterRule struct {
+	Short       string
+	Description string
+}
+
+type Warning struct {
+	Short    string        `json:"short"`
+	Detail   [][]byte      `json:"detail"`
+	Location *parser.Range `json:"location"`
+	Filename string        `json:"filename"`
+	Source   []string      `json:"source"`
+}
+
+func init() {
+	LinterRules = map[string]LinterRule{
+		RuleStageNameCasing: {
+			Short:       "Lowercase Staging Name",
+			Description: "Stage names should be lowercase",
+		},
+		RuleFlagCasing: {
+			Short:       "Lowercase Command Flags",
+			Description: "Flags should be lowercase",
+		},
+		RuleCommandCasing: {
+			Short:       "Inconsistent Command Casing",
+			Description: "Commands should be in consistent casing (all lower or all upper)",
+		},
+	}
+}
+
+func isUpperCase(s string) bool {
+	return s == strings.ToUpper(s)
+}
+
+func isLowerCase(s string) bool {
+	return s == strings.ToLower(s)
+}
+
+func consistentCasing(s string) bool {
+	return isUpperCase(s) || isLowerCase(s)
+}
+
+func ValidateStageNameCasing(cmdArgs []string) bool {
+	if len(cmdArgs) != 3 {
+		return true
+	}
+	stageName := cmdArgs[2]
+	return isLowerCase(stageName)
+}
+
+func ValidateFlagCasing(flags []string) bool {
+	for _, flag := range flags {
+		if !isLowerCase(flag) {
+			return false
+		}
+	}
+	return true
+}
+
+func ValidateCommandCasing(cmd string) bool {
+	return consistentCasing(cmd)
+}
+
+func ValidateNodeLintRules(node *parser.Node) []parser.Warning {
```

Comment:
I think this actual validation logic belongs to the caller side. Atm. we have 2 rules in linter package and 1 in instructions pkg. Better to keep the linter pkg only for the definitions and helpers for triggering warnings rather than include actual build logic over `parser.Node` etc. in here.